[e2e] Can Edit and Delete Product Review

alvarothomas opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce-quality/issues/610

Closes https://github.com/woocommerce/woocommerce-quality/issues/611

The tests focus on editing and deleting product reviews, verifying the visibility of review elements in the admin interface as a merchant.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Code review.
2. Verify if there are any additional steps or assertions that need to be included
3. Run `pnpm test:e2e-pw ./tests/e2e-pw/tests/merchant/product-reviews.spec.js --headed` and make sure that all the tests are passing.
4. CI jobs should be green.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add e2e tests to edit and delete product reviews.

#### Comment

</details>

Commit:
Update product-review

Add test to edit a product review and another test to delete it

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

adimoldovan commented on plugins/woocommerce/changelog/44411-e2e-can-edit-product-review:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: dev
+
+Add e2e tests to edit and delete product reviews.
```
Minor, but there should be a new line at the end of the file.

alvarothomas replied:
Interesting, I used the template on PR creation. It should be fixed now.

adimoldovan replied:
Oh, I see. Then we should probably update the tool that automatically creates the changelog. I'll check it out.

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -60,4 +62,66 @@ test.describe( 'Products > Reviews', () => {
+
+		// Select Trash action, check confirmation prompt and undo
+		await page.getByRole( 'button', { name: 'Trash' } ).click();
+		await page.waitForTimeout( 2000 );
```
It's not a good practice to use waitForTimeout. Instead, you should wait for something to happen in the app, like for an element to be displayed. 
In fact that's exactly what the next line does, it waits for element with the specific message, so this wait is not helping at all, it only adds 2 seconds to the test execution time.

alvarothomas replied:
Thanks for this, I left them all while I was adding steps to the test and forgot to remove them.

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -60,4 +62,66 @@ test.describe( 'Products > Reviews', () => {
+
+		// Select Trash action and delete it permanently
+		await page.getByRole( 'button', { name: 'Trash' } ).click();
+		await page.waitForTimeout( 2000 );
```
Same as above, `waitForTimeout` is discouraged.

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -60,4 +62,66 @@ test.describe( 'Products > Reviews', () => {
+		await page.getByRole( 'button', { name: 'Quick Edit' } ).click();
+		await page.locator( '.wp-editor-area' ).first().fill( updatedReview ); //updatedReview
+		await page.getByRole( 'button', { name: 'Update Comment' } ).click();
+		await page.waitForTimeout( 2000 );
```
Same as the other comment, `waitForTimeout` is discouraged.

adimoldovan reviewed (changes requested):
Hi @alvarothomas, thanks for working on this.

The problem I see with this spec is that all 3 tests are using the same product, and if  they are to run in parallel they will impact each other. The delete test will probably delete the review and the other 2 tests will fail.
Each test should be independent, and have its own test data. I would suggest to either use a `beforeEach`, that will create a product before each test instead of a single product beforeAll, or use a fixture, that will isolate the data for each test. [Here's an example](https://github.com/woocommerce/woocommerce/blob/2628a5238556c66002d49af4f094e3021a1cf646/plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/product-edit-block-editor.spec.js#L22) of a product fixture.

alvarothomas commented:
> Hi @alvarothomas, thanks for working on this.
> 
> The problem I see with this spec is that all 3 tests are using the same product, and if they are to run in parallel they will impact each other. The delete test will probably delete the review and the other 2 tests will fail. Each test should be independent, and have its own test data. I would suggest to either use a `beforeEach`, that will create a product before each test instead of a single product beforeAll, or use a fixture, that will isolate the data for each test. [Here's an example](https://github.com/woocommerce/woocommerce/blob/2628a5238556c66002d49af4f094e3021a1cf646/plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/product-edit-block-editor.spec.js#L22) of a product fixture.

That's a very good point, @adimoldovan . Thank you for suggesting this. The solution of using fixtures seems indeed cleaner and reusable.

Commit:
Added test data in fixture and changelog extra line

Change test data use from constants at the top to test data created in a fixture for each test. Also, added extra line to the changelog file.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Update changelog

Fixed mixed lines on previous commit.

Commit:
Merge branch 'e2e-can-edit-product-review' of https://github.com/woocommerce/woocommerce into e2e-can-edit-product-review

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
+
+		// Create new review, Quick Edit it and save
+		const EditedtimeStamp = Date.now().toString();
+		const updatedReview = `(edited ${ EditedtimeStamp })`;
```
Minor, but `EditedtimeStamp` doesn't follow the naming convention, it should be `editedTimestamp`.
But even better, I suggest removing it completely:

```suggestion
		const updatedReview = `(edited ${ Date.now() })`;
```

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
+		).toBeVisible();
+
+		//Hover over the product review to have 'Edit' displayed
+		await page.hover( '.comment-text' );
```
This will always hover the first review, which may not be the one you want to hover. I think you need to find locators that are linked to the review id, which is included in the id of the table row.

```suggestion
		await page.hover( `#comment-${ testData.review.id }` );
```

alvarothomas replied:
Right, that way it takes the one created in the fixture.

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
+		// Create new review, Quick Edit it and save
+		const EditedtimeStamp = Date.now().toString();
+		const updatedReview = `(edited ${ EditedtimeStamp })`;
+		await page.getByRole( 'button', { name: 'Quick Edit' } ).click();
```
This one fails in case there are more than one reviews available in the table. Similar to my suggestion above in the case of hover, you need to link to the review id.

```suggestion
		await page
			.locator( `#comment-${ testData.review.id }` )
			.getByRole( 'button', { name: 'Quick Edit' } )
			.click();
```

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
+		const commentText = page.locator( '.comment-text' );
+		await commentText.waitFor( { state: 'visible' } );
+		const updatedComment = await commentText.innerText();
+		await expect( updatedComment ).toBe( updatedReview );
```
A single line should do the job here.

```suggestion
		await expect( page.getByText( updatedReview ) ).toBeVisible();
```

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
+		const reviewerName = testData.review.reviewer;
+
+		//Hover over the product review to have 'Edit' displayed
+		await page.hover( '.comment-text' );
```
Similar to the `edit` test, need an unique locators so it doesn't fail if there are more reviews in the table.

```suggestion
		await page.hover( `#comment-${ testData.review.id }` );
```

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
+		await page.hover( '.comment-text' );
+
+		// Select Trash action, check confirmation prompt and undo
+		await page.getByRole( 'button', { name: 'Trash' } ).click();
```
```suggestion
		await page
			.locator( `#comment-${ testData.review.id }` )
			.getByRole( 'button', { name: 'Trash' } )
			.click();
```

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
 		).toBeVisible();
+
+		// Select Trash action and delete it permanently
+		await page.getByRole( 'button', { name: 'Trash' } ).click();
```
```suggestion
		await page
			.locator( `#comment-${ testData.review.id }` )
			.getByRole( 'button', { name: 'Trash' } )
			.click();
```

adimoldovan commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js:
```diff
@@ -1,63 +1,142 @@
+
+		// Assert that the message is present, indicating an empty comment list
+		const noReviewsFound = await page.$( 'tr.no-items' );
+		expect( noReviewsFound ).not.toBeNull();
```
This check again assumes that this review will be the only one available and the table will be empty after deletion, which is not something the test can control.

I would suggest using two checks instead:
1. Switch to "Trash" and validate that the review is in there.
2. Access the review with a direct link (e.g. `wp-admin/comment.php?action=editcomment&c=${testData.review.id}`) and check the displayed message: `This comment is in the Trash. Please move it out of the Trash if you want to edit it.`

alvarothomas replied:
Done, thank you so much for this idea. It makes a lot of sense to check that instead. Also, the direct link tip helps a lot 😄

adimoldovan reviewed (changes requested):
The fixture setup looks great @alvarothomas.
I checked the tests again and found some locators that should be updated to be unique to the review, otherwise the tests will fail if there are other reviews in the table besides the one in your test (which is very possible). See my suggested edits.
Also, the delete test needs another assertion, for the same reason. See my comments below.

I'm not sure about the scope of the task (if quick edit was the scope), but I'm wondering if editing the review via the review edit page should also be covered.
The same though with bulk delete.

Commit:
Update plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js

Remove edited timestamp

Co-authored-by: Adrian Moldovan <3854374+adimoldovan@users.noreply.github.com>

Commit:
Update plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js

Update hover action to review id.

Co-authored-by: Adrian Moldovan <3854374+adimoldovan@users.noreply.github.com>

Commit:
Update plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js

Change Quick Edit button to reach review identifier

Co-authored-by: Adrian Moldovan <3854374+adimoldovan@users.noreply.github.com>

Commit:
Update plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js

Replace updated review check to single line assert

Co-authored-by: Adrian Moldovan <3854374+adimoldovan@users.noreply.github.com>

Commit:
Update plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js

Hover to edit on review identifier

Co-authored-by: Adrian Moldovan <3854374+adimoldovan@users.noreply.github.com>

Commit:
Update plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js

Point to trash a review to its identifier.

Co-authored-by: Adrian Moldovan <3854374+adimoldovan@users.noreply.github.com>

Commit:
Update plugins/woocommerce/tests/e2e-pw/tests/merchant/product-reviews.spec.js

Point to trash a review to its identifier.

Co-authored-by: Adrian Moldovan <3854374+adimoldovan@users.noreply.github.com>

Commit:
Added Review Edit test

Added Review Edit test apart from Quick Edit test. Included assertions to verify that the review has been edited and can be seen both in the reviews list and the shop's page.

adimoldovan reviewed (approved):
This looks good! Thank you @alvarothomas.
